### PR TITLE
Adding unit tests to cluster_restart

### PR DESCRIPTION
**Pull request by udnay**
Attempt at adding some unit tests to the operator using the fake client and stubbed resource objects. I would like to get some comments on if the testing approach makes sense to quickly test the logic of the operator.

### CODE REVIEW

File: pkg/actor/cluster_restart_test.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+	sts.Status.Replicas = stsReplicas
+	sts.Status.ReadyReplicas = stsReplicas
+
+	cltSet.Tracker().Update(schema.GroupVersionResource{
+		Group:    "apps",
+		Version:  "v1",
+		Resource: "statefulset",
+	}, &sts, sts.Namespace)
+	require.NoError(t, cr.fullClusterRestart(context.TODO(), &sts, Log, cltSet))
```
Comment (udnay) [inline pkg/actor/cluster_restart_test.go:166]: @chrisseto this is the call where I'd like to make sure that the restart doesn't complete if all replicas are not ready yet and where I wanted to add in some mocks to add replicas after an initial failure of the check.

File: pkg/actor/cluster_restart.go [hunk 2]
```diff
@@ -147,6 +147,13 @@ func (r *clusterRestart) Act(ctx context.Context, cluster *resource.Cluster) err
 	return nil
 }
 
+func statefulSetReplicasAvailable(status *v1.StatefulSetStatus) error {
```
Comment (chrislovecnm) [inline pkg/actor/cluster_restart.go:150]: We should externalize this function, we use this all over the code base

File: pkg/actor/cluster_restart_test.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+func createStatefulSet(stsReplicas int32) appsv1.StatefulSet {
```
Comment (chrislovecnm) [inline pkg/actor/cluster_restart_test.go:169]: We already have a builder, we should reuse it if we cab

File: pkg/scale/cockroach_statefulset_test.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	require.NoError(t, IsStatefulSetReadyToServe(context.TODO(), cltSet, "crdb", "crdb-sts", stsReplicas))
+
+	// Err if there are more replicas running than in the spec
+	sts.Status.Replicas = stsReplicas * 2
+	cltSet.Tracker().Update(sts.GroupVersionKind().GroupVersion().WithResource("statefulsets"), &sts, "crdb")
+	require.Error(t, IsStatefulSetReadyToServe(context.TODO(), cltSet, "crdb", "crdb-sts", stsReplicas))
+
+}
+
+func statefulSet(stsReplicas int32) appsv1.StatefulSet {
```
Comment (chrislovecnm) [inline pkg/scale/cockroach_statefulset_test.go:121]: If we can reuse this code we need it in the test utils pkg

File: pkg/actor/cluster_restart_test.go (hunk 3 above)
Comment (udnay) [reply pkg/actor/cluster_restart_test.go:240]: The builder is tied to `e2e` or at least was, can we do that in a follow up PR?

File: pkg/scale/cockroach_statefulset_test.go (hunk 4 above)
Comment (udnay) [reply pkg/scale/cockroach_statefulset_test.go:134]: I agree, I can refactor in a followup PR but don't want to keep this initial template to be out.

**Review by keith-mcclellan — DISMISSED** [review]

File: pkg/actor/cluster_restart_test.go [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+			},
+		}
+		require.False(t, cra.Handles(conditions))
+	})
+
+}
+
+func TestClusterReadyForRestart(t *testing.T) {
+	ssStatus := appsv1.StatefulSetStatus{
+		Replicas:        2,
```
[the remark this answers is not in the document]
Comment (udnay) [reply pkg/actor/cluster_restart_test.go:126]: We're not testing the DB though. It's probably an anti-pattern, you're right. I can change these numbers.

File: pkg/scale/cockroach_statefulset_test.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+					"app.kubernetes.io/name":      "crdb-sts",
+				},
+				MatchExpressions: nil,
+			},
+			Replicas: &stsReplicas,
+		},
+	}
+}
+
+func addPodsToStatefulSet(stsReplicas int32, sts appsv1.StatefulSet, cltSet *fakeclient.Clientset) {
```
[the remark this answers is not in the document]
Comment (udnay) [reply pkg/scale/cockroach_statefulset_test.go:159]: Ya, just copy paste for now, will refactor in a follow up PR, is that acceptable?

File: pkg/actor/cluster_restart_test.go (hunk 1 above)
Comment (udnay) [reply pkg/actor/cluster_restart_test.go:184]: This is testing to make sure that `fullClusterRestart` does not return an error. These tests are focused on the behaviour of the operator *not* the underlying DB. e.g. If the operator is in this state what are the different actions it takes and then what state is the operator in after that. Any calls to the DB like `HealthChecker` are mocked.

Comment (keith-mcclellan) [reply pkg/actor/cluster_restart_test.go:184]: 👍

File: pkg/scale/cockroach_statefulset_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (keith-mcclellan) [reply pkg/scale/cockroach_statefulset_test.go:159]: As long as we have an issue to track before we close this out

[the remark this answers is not in the document]
Comment (chrislovecnm) [reply pkg/scale/cockroach_statefulset_test.go:159]: @udnay do you have the issues open?

File: pkg/actor/cluster_restart_test.go [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+	require.NoError(t, cr.fullClusterRestart(context.TODO(), &sts, Log, cltSet))
+}
+
+func TestRollingClusterRestart(t *testing.T) {
+	// Setup fake client
+	builder := fake.NewClientBuilder()
+
+	client := builder.Build()
+
+	cr := newClusterRestart(nil, client, nil).(*clusterRestart)
```
Comment (chrislovecnm) [inline pkg/actor/cluster_restart_test.go:193]: Would we ever pass in nils?

File: pkg/scale/cockroach_statefulset_test.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	fakeclient "k8s.io/client-go/kubernetes/fake"
+	clienttesting "k8s.io/client-go/testing"
+)
+
+func TestStatefulSetIsRunning(t *testing.T) {
```
Comment (chrislovecnm) [inline pkg/scale/cockroach_statefulset_test.go:34]: What about if the sts is not found?

File: pkg/scale/cockroach_statefulset_test.go (hunk 4 above)
Comment (chrislovecnm) [reply pkg/scale/cockroach_statefulset_test.go:134]: Please file an issue

**Review by chrislovecnm — COMMENTED** [review]
Looking good.  I have some questions, and I would like to get this merged!

File: pkg/scale/cockroach_statefulset_test.go (hunk 4 above)
Comment (udnay) [reply pkg/scale/cockroach_statefulset_test.go:134]: https://github.com/cockroachdb/cockroach-operator/issues/562 done

File: pkg/scale/cockroach_statefulset_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (udnay) [reply pkg/scale/cockroach_statefulset_test.go:159]: https://github.com/cockroachdb/cockroach-operator/issues/562 yes

File: pkg/actor/cluster_restart_test.go (hunk 7 above)
Comment (udnay) [reply pkg/actor/cluster_restart_test.go:193]: Nope, I just need the minimal things to get a `ClusterRestart` actor to test the type methods

File: pkg/actor/cluster_restart_test.go [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+			},
+			{
+				Type:               api.CrdbVersionChecked,
+				Status:             metav1.ConditionFalse,
+				LastTransitionTime: metav1.Now(),
+			},
+		}
+		require.False(t, cra.Handles(conditions))
+
+		// Initialized condition is unknown or missing
```
[the remark this answers is not in the document]
Comment (udnay) [reply pkg/actor/cluster_restart_test.go:97]: Looking at the code I think they amount to the same thing?

File: pkg/scale/cockroach_statefulset_test.go (hunk 8 above)
Comment (udnay) [reply pkg/scale/cockroach_statefulset_test.go:34]: Added

File: pkg/actor/cluster_restart_test.go [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+func (hc *HealthCheckerTest) Probe(ctx context.Context, l logr.Logger, logSuffix string, partition int) error {
+	return nil
+}
+
+func TestClusterRestartHandles(t *testing.T) {
+	cra := clusterRestart{}
+
+	var conditions []api.ClusterCondition
+
+	t.Run("Handle func passes", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (udnay) [reply pkg/actor/cluster_restart_test.go:49]: I added some comments but I don't know all the conditions to be honest. Basically we can execute a cluster restart whenever we have checked the CRDBVersion

File: pkg/actor/cluster_restart_test.go (hunk 3 above)
Comment (udnay) [reply pkg/actor/cluster_restart_test.go:242]: You want to use the builder for building something other than the CRDB Cluster CRD?

**Review by chrislovecnm — APPROVED** [review]
LGTM

**Pull request merged**